core/commands/dag/dag.go

```diff
@@ -241,3 +248,103 @@ var DagResolveCmd = &cmds.Command{
 	},
 	Type: ResolveOutput{},
 }
+
+var DagExportCmd = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "Streams the selected DAG as a .car stream on stdout.",
+		ShortDescription: `
+'ipfs dag export' fetches a dag and streams it out as a well-formed .car file.
+Note that at present only single root selections / .car files are supported.
+The output of blocks happens in strict DAG-traversal, first-seen, order.
+`,
+	},
+	Arguments: []cmds.Argument{
+		cmds.StringArg("root", true, false, "Expression evaluting to a single root of a dag to export").EnableStdin(),
+	},
+	Run: func(req *cmds.Request, res cmds.ResponseEmitter, env cmds.Environment) error {
+
+		c, err := cid.Decode(req.Arguments[0])
+		if err != nil {
+			return fmt.Errorf(
+				"unable to parse root specification (currently only bare CIDs are supported): %s",
+				err,
+			)
+		}
+
+		// The current interface of go-car is rather suboptimal as it
+		// only takes a blockstore, instead of accepting a dagservice,
+		// and leveraging parallel-fetch capabilities
+		// https://github.com/ipld/go-car/issues/27
+		//
+		// Until the above is fixed, pre-warm the blockstore before doing
+		// anything else. We explicitly *DO NOT* take a lock during this
+		// operation: even if we lose some of the blocks we just received
+		// due to a conflicting GC: we will just re-retrieve anything we
+		// potentially lost when the car is being streamed out
+		node, err := cmdenv.GetNode(env)
+		if err != nil {
+			return err
+		}
+
+		if err := mdag.FetchGraph(req.Context, c, node.DAG); err != nil {
+			if !node.IsOnline {
+				err = fmt.Errorf("%s (currently offline, perhaps retry after attaching to the network)", err)
+			}
+			return err
+		}
+
+		// Code disabled until descent-issue in go-ipld-prime is fixed
+		//
+		// The second part of the above - make a super-thin wrapper around
+		// a blockservice session, translating Session.GetBlock() to Blockstore.Get()
+		//
+		// sess := blockservice.NewSession(
+		// 	req.Context,
+		// 	node.Blocks,
+		// )
+		// var wrapper getBlockFromSessionWrapper = func(c cid.Cid) (blk.Block, error) {
+		// 	return sess.GetBlock(req.Context, c)
+		// }
+		// sb := gipselectorbuilder.NewSelectorSpecBuilder(gipfree.NodeBuilder())
+		// car := gocar.NewSelectiveCar(
+		// 	req.Context,
+		// 	&wrapper,
+		// 	[]gocar.Dag{gocar.Dag{
+		// 		Root: c,
+		// 		Selector: sb.ExploreRecursive(
+		// 			gipselector.RecursionLimitNone(),
+		// 			sb.ExploreAll(sb.ExploreRecursiveEdge()),
+		// 		).Node(),
+		// 	}},
+		// )
+
+		pipeR, pipeW := io.Pipe()
+
+		errCh := make(chan error, 2) // we only report the 1st error
+		go func() {
+			defer func() {
+				if err := pipeW.Close(); err != nil {
+					errCh <- fmt.Errorf("stream flush failed: %s", err)
+				}
+				close(errCh)
+			}()
+
+			//if err := car.Write(pipeW); err != nil {
+			if err := gocar.WriteCar(
+				req.Context,
+				node.DAG,
```

Comment:
Ideally, we'd call `merkledag.NewSession(ctx, node.DAG)`. But we'll have to change go-car to accept an `ipld-format.NodeGetter` instead of requiring a full `DAGService`.

Reply:
This is not applicable - we will switch to the selector-based interface the moment go-ipld-prime is fixed ( the interface in `carobj.Write` is different 😿 )

Reply:
But that's not ready yet and may not be ready for a while. This fix is easy: https://github.com/ipld/go-car/pull/28

Reply:
Fair enough, I am still not in a spot where I can tell when it is ok to change an interface other teams use...
Updated to use your branch!